### ISSUE TITLE
Non contiguous target prevents learning using F.cross_entropy on GPU

### ISSUE THREAD

**Issue by GabrielBena** [pytorch/pytorch#72120]
### 🐛 Describe the bug

When using a simple, modified version of MNIST dataset returning pairs of input, I encountered a weird bug preventing training when using the cross_entropy loss function. I needed a double-digit version on MNIST to train of different possibility of targets (sum, 100 class etc...). However, on a simple case scenario optimization failed. It had to do with my indexing of an 2 column target AFTER sending it to the GPU. That resulted in a non contiguous tensor (stride 2). Any operation on the target made training possible again, but the in appearance the simplest of all tasks (predicting 1st digit only) seemed not to train. Actually, the loss actually goes back up when training with non-contiguous version of the target, even on a properly trained model.

Solution : make target contiguous again using .contiguous(), or for that matter any operation such as target = target + 1 - 1

This is my first time reporting a bug so hopes it helps and is not to messy !! All the best 

Python : 

`
    from torch.utils.data import Dataset, DataLoader
    import torch
    from torchvision import transforms, datasets
    import torch.nn as nn
    import torch.nn.functional as F
    from tqdm.notebook import tqdm
    import os

    use_cuda = True
    device = torch.device("cuda" if use_cuda else "cpu")
    batch_size = 128
    train_kwargs = {'batch_size': batch_size, 'shuffle': True}
    if use_cuda:
        cuda_kwargs = {'num_workers': 0,
                    'pin_memory': True
                    }
        train_kwargs.update(cuda_kwargs)

    root = os.path.expanduser("~/Code/Data/MNIST")
    transformations= transforms.Compose(
        [transforms.ToTensor(), 
        transforms.Normalize((0.1307,), (0.3081,))
        ])

    class DoubleMNIST(Dataset) :
        """
        Double Digits MNIST dataset    
        """
        
        def __init__(self, root, train=True, transforms=None):
            super().__init__()

            self.transform=transforms
            dataset = datasets.MNIST(root, train=train, download=False,
                    transform=self.transform)
            self.mnist_dataset = dataset
            self.secondary_index = torch.randperm(len(self.mnist_dataset))

        def __getitem__(self, index):
            index_1= index
            index_2 = self.secondary_index[index]

            digit_1, target_1 = self.mnist_dataset[index_1]
            digit_2, target_2 = self.mnist_dataset[index_2]

            digits = torch.cat([digit_1, digit_2], axis = 0)
            targets = [target_1, target_2]

            return digits, torch.tensor(targets)

        def __len__(self) : 
            return len(self.mnist_dataset)

    double_dataset = DoubleMNIST(root, True, transformations)
    loader = DataLoader(double_dataset, **train_kwargs)

    model = nn.Sequential(nn.Linear(784, 100), nn.ReLU(), nn.Linear(100, 10)).to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

    pbar = tqdm(range(1))

    model.train()     
    for epoch in pbar : 
        for batch_idx, batch in enumerate(loader):

            data, target = batch[0].to(device), batch[1].to(device)
            data, target = data[:, 0].flatten(start_dim=-2), target[:, 0]

            #print(target.is_contiguous())
            #target = target.contiguous() #solves the issue

            optimizer.zero_grad()
            output = model(data)

            assert target.shape == output.shape[:-1], print(target.shape, output.shape)

            loss = F.cross_entropy(output, target, reduction='mean')
                
            pred = output.argmax(dim=-1)  # get the index of the max log-probability
            correct = pred.eq(target.view_as(pred)).sum().item()
            acc = correct/target.numel()
            
            loss.backward()
            optimizer.step()

            desc = str('Train Epoch: {} [{}/{} ({:.0f}%)] Loss: {:.3f}, Acc:{}'.format(
                    epoch, batch_idx * loader.batch_size, len(loader.dataset),
                    100. * batch_idx / len(loader), loss.item(), acc))

            pbar.set_description(desc)

`

### Versions

Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-3.10.0-862.14.4.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 465.19.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] functorch==0.0.1a0+6107f49
[pip3] numpy==1.20.3
[pip3] torch==1.10.0
[pip3] torchaudio==0.10.0
[pip3] torchvision==0.11.1
[pip3] torchviz==0.0.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.20.3                   pypi_0    pypi
[conda] numpy-base                1.20.1           py39h7d8b39e_0  
[conda] pytorch                   1.10.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.10.0               py39_cu113    pytorch
[conda] torchvision               0.11.1               py39_cu113    pytorch
[conda] torchviz                  0.0.2                    pypi_0    pypi

**Comment by ngimel** [comment]
Fixed in 1.10.1, please reopen if still exists.
